statistics/sample_test.go

```diff
@@ -232,77 +85,117 @@ func (s *testSampleSuite) TestDistributedWeightedSampling(c *C) {
 				Rng:             rand.New(rand.NewSource(time.Now().UnixNano())),
 			}
 			collector, err := builder.Collect()
-			c.Assert(err, IsNil)
-			rootRowCollector.MergeCollector(collector)
-			c.Assert(sets[i].Close(), IsNil)
+			require.NoError(t, err)
+			for i := 0; i < collector.MaxSampleSize; i++ {
+				a := collector.Samples[i].Columns[0].GetInt64()
+				itemCnt[a]++
+			}
+			require.Nil(t, rs.Close())
 		}
-		for _, sample := range rootRowCollector.Samples {
-			itemCnt[sample.Columns[0].GetInt64()]++
+		expFrequency := float64(sampleNum) * float64(loopCnt) / float64(rowNum)
+		for _, cnt := range itemCnt {
+			require.InEpsilonf(t, expFrequency, float64(cnt), 0.5, "The frequency %v is exceed the Chernoff Bound", cnt)
 		}
 	}
-	expFrequency := float64(sampleNum) * float64(loopCnt) / float64(rowNum)
-	delta := 0.5
-	for _, cnt := range itemCnt {
-		if float64(cnt) < expFrequency/(1+delta) || float64(cnt) > expFrequency*(1+delta) {
-			c.Assert(false, IsTrue, Commentf("the frequency %v is exceed the Chernoff Bound", cnt))
+}
+
+func SubTestDistributedWeightedSampling() func(*testing.T) {
+	return func(t *testing.T) {
+		t.Parallel()
+		sampleNum := int64(10)
+		rowNum := 100
+		loopCnt := 1500
+		batch := 5
+		sets := recordSetForDistributedSamplingTest(rowNum, batch)
+		sc := mock.NewContext().GetSessionVars().StmtCtx
+		// The loop which is commented out is used for stability test.
+		// This test can run 800 times in a row without any failure.
+		// for x := 0; x < 800; x++ {
+		itemCnt := make([]int, rowNum)
+		for loopI := 1; loopI < loopCnt; loopI++ {
+			rootRowCollector := NewReservoirRowSampleCollector(int(sampleNum), 1)
+			rootRowCollector.FMSketches = append(rootRowCollector.FMSketches, NewFMSketch(1000))
+			for i := 0; i < batch; i++ {
+				builder := &ReservoirRowSampleBuilder{
+					Sc:              sc,
+					RecordSet:       sets[i],
+					ColsFieldType:   []*types.FieldType{types.NewFieldType(mysql.TypeLonglong)},
+					Collators:       make([]collate.Collator, 1),
+					ColGroups:       nil,
+					MaxSampleSize:   int(sampleNum),
+					MaxFMSketchSize: 1000,
+					Rng:             rand.New(rand.NewSource(time.Now().UnixNano())),
+				}
+				collector, err := builder.Collect()
+				require.NoError(t, err)
+				rootRowCollector.MergeCollector(collector)
+				require.Nil(t, sets[i].Close())
+			}
+			for _, sample := range rootRowCollector.Samples {
+				itemCnt[sample.Columns[0].GetInt64()]++
+			}
+		}
+		expFrequency := float64(sampleNum) * float64(loopCnt) / float64(rowNum)
+		for _, cnt := range itemCnt {
+			require.InEpsilonf(t, expFrequency, float64(cnt), 0.5, "The frequency %v is exceed the Chernoff Bound", cnt)
```

Comment:
They do not seem to be completely equivalent.

[InEpsilon](https://github.com/stretchr/testify/blob/ab6dc3262822ed562480c19876b0257ace761e3e/assert/assertions.go#L1259-L1299)(expFrequency, float64(cnt), 0.5) ==>
`-0.5` < (float64(cnt) - expFrequency) / expFrequency < `0.5`

float64(cnt) > expFrequency/(1+delta) && float64(cnt) < expFrequency*(1+delta) ==>
float64(cnt)/expFrequency > 1/(1+delta) && float64(cnt)/expFrequency < 1+delta ==>
float64(cnt)/expFrequency - 1 > -delta/(1+delta) && float64(cnt)/expFrequency - 1 < delta ==>
-delta/(1+delta) < float64(cnt)/expFrequency - 1 < delta ==>
-delta/(1+delta) < (float64(cnt) - expFrequency)/expFrequency < delta ==>
`-0.333` < (float64(cnt) - expFrequency)/expFrequency < `0.5`

Reply:
yes, you are right.
use `InEpsilon` here seems could not make the logic look more clear, reverted.